rhizome/kubernetes/bin/ubicni

```diff
@@ -0,0 +1,295 @@
+    end
+  end
+
+  def handle_add
+    subnet_ula_ipv6 = @stdin_data["ranges"]["subnet_ula_ipv6"]
+    subnet_ipv6 = @stdin_data["ranges"]["subnet_ipv6"]
+    subnet_ipv4 = @stdin_data["ranges"]["subnet_ipv4"]
+
+    container_id = ENV["CNI_CONTAINERID"]
+    cni_netns = ENV["CNI_NETNS"].sub("/var/run/netns/", "")
+
+    inner_mac = gen_mac.shellescape
+    inner_link_local = mac_to_ipv6_link_local(inner_mac)
+    inner_ifname = ENV["CNI_IFNAME"]
+
+    outer_mac = gen_mac.shellescape
+    outer_link_local = mac_to_ipv6_link_local(outer_mac)
+    outer_ifname = "veth_#{container_id[0, 8]}"
+
+    FileUtils.mkdir_p("/etc/netns/#{cni_netns}")
+    dns_config = <<~EOF
+nameserver 10.96.0.10
+search default.svc.cluster.local svc.cluster.local cluster.local
+options ndots:5
+    EOF
+    File.write("/etc/netns/#{cni_netns}/resolv.conf", dns_config)
+
+    r "ip link add #{outer_ifname} addr #{outer_mac} type veth peer name #{inner_ifname} addr #{inner_mac} netns #{cni_netns}"
+
+    container_ipv6 = setup_ipv6(subnet_ipv6, inner_link_local, outer_link_local, cni_netns, inner_ifname, outer_ifname, setup_default_route: true)
+    container_ula_ipv6 = setup_ipv6(subnet_ula_ipv6, inner_link_local, outer_link_local, cni_netns, inner_ifname, outer_ifname)
+    ipv4_ips = setup_ipv4(subnet_ipv4, cni_netns, inner_ifname, outer_ifname)
+    ipv4_container_ip, ipv4_gateway_ip = ipv4_ips[:container_ip], ipv4_ips[:gateway_ip]
+
+    @ipam_store["allocated_ips"][container_id] = [ipv4_container_ip.to_s, ipv4_gateway_ip.to_s, container_ula_ipv6.to_s, container_ipv6.to_s]
+    save_state
+
+    response = {
+      cniVersion: "1.0.0",
+      interfaces: [
+        {
+          name: inner_ifname,
+          mac: inner_mac,
+          sandbox: "/var/run/netns/#{cni_netns}"
+        }
+      ],
+      ips: [
+        {
+          address: "#{ipv4_container_ip}/#{ipv4_container_ip.prefix}",
+          gateway: ipv4_gateway_ip.to_s,
+          interface: 0
+        },
+        {
+          address: "#{container_ula_ipv6}/#{container_ula_ipv6.prefix}",
+          gateway: outer_link_local,
+          interface: 0
+        },
+        {
+          address: "#{container_ipv6}/#{container_ipv6.prefix}",
+          gateway: outer_link_local,
+          interface: 0
+        }
+      ],
+      routes: [
+        {
+          dst: "0.0.0.0/0"
+        }
+      ],
+      dns: {
+        nameservers: ["10.96.0.10"],
+        search: ["default.svc.cluster.local", "svc.cluster.local", "cluster.local"],
+        options: ["ndots:5"]
+      }
+    }
+    LOG.info "add response: #{JSON.generate(response)}"
+    puts JSON.generate(response)
+  end
+
+  def setup_ipv6(subnet, inner_link_local, outer_link_local, cni_netns, inner_ifname, outer_ifname, setup_default_route: false)
+    container_ip = find_random_available_ip(IPAddr.new(subnet))
+
+    r "ip -6 -n #{cni_netns} addr add #{container_ip}/#{container_ip.prefix} dev #{inner_ifname}"
```

Comment:
you may want to use `addr replace` instead of `addr add` it helps with idempotancy.